library/waku_thread/inter_thread_communication/node_lifecycle_request.nim

```diff
@@ -6,32 +6,121 @@ import
+  ../../../waku/waku_core,
   ../../../waku/node/waku_node,
+  ../../../waku/node/builder,
+  ../../../waku/node/config,
+  ../../../waku/waku_relay/protocol,
+  ../../events/[json_error_event,json_message_event,json_base_event],
+  ../config,
   ./request
 
 type
   NodeLifecycleMsgType* = enum
+    CREATE_NODE
     START_NODE
     STOP_NODE
 
 type
   NodeLifecycleRequest* = ref object of InterThreadRequest
     operation: NodeLifecycleMsgType
+    configJson: cstring ## Only used in 'CREATE_NODE' operation
 
 proc new*(T: type NodeLifecycleRequest,
-          op: NodeLifecycleMsgType): T =
+          op: NodeLifecycleMsgType,
+          configJson: cstring = ""): T =
 
-  return NodeLifecycleRequest(operation: op)
+  return NodeLifecycleRequest(operation: op, configJson: configJson)
+
+proc createNode(configJson: cstring): Result[WakuNode, string] =
+  var privateKey: PrivateKey
+  var netConfig = NetConfig.init(ValidIpAddress.init("127.0.0.1"),
+                                 Port(60000'u16)).value
+  var relay: bool
+  var topics = @[""]
+  var jsonResp: JsonEvent
+
+  if not parseConfig($configJson,
+                     privateKey,
+                     netConfig,
+                     relay,
+                     topics,
+                     jsonResp):
+    return err($jsonResp)
+
+  var enrBuilder = EnrBuilder.init(privateKey)
+
+  enrBuilder.withIpAddressAndPorts(
+    netConfig.enrIp,
+    netConfig.enrPort,
+    netConfig.discv5UdpPort
+  )
+
+  if netConfig.wakuFlags.isSome():
+    enrBuilder.withWakuCapabilities(netConfig.wakuFlags.get())
+
+  enrBuilder.withMultiaddrs(netConfig.enrMultiaddrs)
+
+  let addShardedTopics = enrBuilder.withShardedTopics(topics)
+  if addShardedTopics.isErr():
+    let msg = "Error setting shared topics: " & $addShardedTopics.error
+    return err($JsonErrorEvent.new(msg))
+
+  let recordRes = enrBuilder.build()
+  let record =
+    if recordRes.isErr():
+      let msg = "Error building enr record: " & $recordRes.error
+      return err($JsonErrorEvent.new(msg))
+
+    else: recordRes.get()
+
+  ## TODO: make the next const configurable from 'configJson'.
+  const MAX_CONNECTIONS = 50.int
+
+  var builder = WakuNodeBuilder.init()
+  builder.withRng(crypto.newRng())
+  builder.withNodeKey(privateKey)
+  builder.withRecord(record)
+  builder.withNetworkConfiguration(netConfig)
+  builder.withSwitchConfiguration(
+    maxConnections = some(MAX_CONNECTIONS)
+  )
+
+  let wakuNodeRes = builder.build()
+  if wakuNodeRes.isErr():
+    let errorMsg = "failed to create waku node instance: " & wakuNodeRes.error
+    return err($JsonErrorEvent.new(errorMsg))
+
+  var newNode = wakuNodeRes.get()
+
+  if relay:
+    waitFor newNode.mountRelay()
+    newNode.peerManager.start()
+
+  return ok(newNode)
 
 method process*(self: NodeLifecycleRequest,
-                node: WakuNode): Future[Result[string, string]] {.async.} =
+                node: ptr WakuNode): Future[Result[string, string]] {.async.} =
```

Comment:
`WakuNode` is a `ref object` - the previous code without the `ptr` was correct - a consequence is that `process` is not a "thread-safe" function - it can only be called from the thread that created the `WakuNode` instance (ie the processing loop below)

Reply:
ah, I see now - this looks as if you wanted to return the node instance and `async` prevents `var` from being used - ok, `ptr` can be used as a workaround for that limitation as long as you keep in mind that a ptr might become dangling - this is a bit of a general problem here and for this PR it's an ok tradeoff to improve the status quo - later, I would probably look for solutions that avoid this quirk

Reply:
Well, the idea of using `ptr` was to alter the node instance: https://github.com/waku-org/nwaku/blob/f29105b6fcb617b975a9815d9c14db82d2fc1148/library/waku_thread/waku_thread.nim#L48

This instance will live enough for all the calls to any `process` proc and will always be owned by the _Waku Thread_.